evm_loader/lib/src/types/account_cache.rs

```diff
@@ -0,0 +1,267 @@
+// use crate::tracing::tracers::state_diff::Account;
+use crate::rpc::Rpc;
+use async_trait::async_trait;
+// use async_trait::async_trait;
+use bincode::deserialize;
+use futures::future::join_all;
+use solana_client::client_error::Result as ClientResult;
+use solana_sdk::{
+    account::Account,
+    bpf_loader_upgradeable::UpgradeableLoaderState,
+
+    // account_utils::StateMut,
+    clock::{Slot, UnixTimestamp},
+    pubkey::Pubkey,
+};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::RwLock;
+
+use bincode::serialize;
+use tokio::sync::OnceCell;
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct KeyAccountCache {
+    addr: Pubkey,
+    slot: u64,
+}
+
+type AccCache = HashMap<KeyAccountCache, Account>;
+type ProtectedAppCache = RwLock<AccCache>;
+
+static LOCAL_CONFIG: OnceCell<ProtectedAppCache> = OnceCell::const_new();
+
+async fn acc_hash_get_instance() -> &'static ProtectedAppCache {
+    LOCAL_CONFIG
+        .get_or_init(|| async {
+            let map = HashMap::new();
+
+            RwLock::new(map)
+        })
+        .await
+}
+
+async fn acc_hash_get(addr: Pubkey, slot: u64) -> Option<Account> {
+    let val = KeyAccountCache { addr, slot };
+    acc_hash_get_instance()
+        .await
+        .read()
+        .expect("acc_hash_get_instance poisoned")
+        .get(&val)
+        .cloned()
+}
+
+async fn acc_hash_add(addr: Pubkey, slot: u64, acc: Account) {
+    let val = KeyAccountCache { addr, slot };
+    acc_hash_get_instance()
+        .await
+        .write()
+        .expect("PANIC, no nable")
+        .insert(val, acc);
+}
+
+fn get_programdata_slot_from_account(acc: &Account) -> u64 {
+    //probably will not serrialize.
+    match deserialize::<UpgradeableLoaderState>(&acc.data) {
+        Ok(UpgradeableLoaderState::ProgramData { slot, .. }) => slot,
+        Ok(_) => {
+            panic!("Account is not of type `ProgramData`.");
+        }
+        Err(e) => {
+            eprintln!("Error occurred: {e:?}");
+            panic!("Failed to deserialize account data.");
+        }
+    }
+}
+
+pub async fn acc_hash_get_values_by_keys(
+    programdata_keys: &Vec<Pubkey>,
+    rpc: &impl Rpc,
+) -> ClientResult<Vec<Option<solana_sdk::account::Account>>> {
+    let mut future_requests = Vec::new();
+    let mut answer = Vec::new();
+
+    for key in programdata_keys {
+        future_requests.push(rpc.get_account_slice(
+            key,
+            0,
+            UpgradeableLoaderState::size_of_programdata_metadata(),
+        ));
+        //future_requests.push(rpc.get_account_slice(key, 0, 512));
+    }
+
+    assert!(
+        programdata_keys.len() == future_requests.len(),
+        "programdata_keys.size()!=future_requests.size()"
+    );
+    let results = join_all(future_requests).await;
+
+    for (i, result) in results.iter().enumerate() {
```

Comment:
It's better to zip two vectors with 
`for (result, key) in results.iter().zip(programdata_keys)`

Reply:
Done!